nomt/src/beatree/ops/bit_ops.rs

```diff
@@ -35,25 +38,124 @@ pub fn separator_len(key: &Key) -> usize {
+        0 => 0,
+        len if prefix_end_bit_offset == 0 => len,
+        // overlap between the end of the prefix and the beginning of the separator
+        len => len - 1,
+    };
+
+    enum Shift {
+        Left(usize),                          // amount
+        Right(usize, Option<u8>, Option<u8>), // amount, prev_remainder, curr_remainder
+    }
+
+    let mut shift = match prefix_end_bit_offset as isize - separator_bit_init as isize {
+        0 => None,
+        shift if shift < 0 => Some(Shift::Left(shift.abs() as usize)),
+        shift => Some(Shift::Right(shift as usize, None, None)),
+    };
+
+    // chunk is an 8-byte slice of the separator which will be cast to a u64 to simplify shifting
+    let n_chunks = separator_bytes.len() / 8;
+    for chunk_index in 0..n_chunks {
+        let chunk_start = chunk_index * 8;
+
+        if let Some(Shift::Right(amount, _prev_remainder, curr_remainder)) = &mut shift {
+            // store bits that will be covered by the right shift
+            let mask = (1 << *amount) - 1;
+            let bits = separator_bytes[chunk_start + 7] & mask;
+            *curr_remainder = Some(bits << (8 - *amount));
         }
-        None => {
-            key.view_bits_mut::<Msb0>()[..separator.len()].copy_from_bitslice(separator);
+
+        let mut chunk = u64::from_be_bytes(
+            separator_bytes[chunk_start..chunk_start + 8]
+                .try_into()
+                .unwrap(),
+        );
+
+        if chunk_index == 0 {
+            // first chunk will probably have garbage at the beginning of the first byte
+            let mask_shift = (7 - separator.1) as u32 + 1 + (8 * 7);
+            let mask = 1u64
+                .checked_shl(mask_shift)
+                .map(|m| m - 1)
+                .unwrap_or(u64::MAX);
+
+            chunk &= mask;
         }
+
+        if chunk_index == n_chunks - 1 {
+            // last chunk will probably have garbage at end
+            let unused_last_bits = separator_bit_init + separator_bit_len - ((n_chunks - 1) * 64);
+            let mask = 1u64
+                .checked_shl(64 - unused_last_bits as u32)
+                .map(|m| !(m - 1))
+                .unwrap_or(0);
+
+            chunk &= mask;
+        }
+
+        match &mut shift {
+            Some(Shift::Left(amount)) => chunk <<= *amount,
+            Some(Shift::Right(amount, _, _)) => chunk >>= *amount,
+            _ => (),
+        };
+
+        let mut chunk_shifted = chunk.to_be_bytes();
+
+        // move bits remainder between chunk bounderies
+        match &mut shift {
+            Some(Shift::Left(amount)) if chunk_index < n_chunks - 1 => {
+                let mask = !(1 << (8 - *amount) - 1);
+                let remainder_bits = (separator_bytes[(chunk_index + 1) * 8] & mask) >> *amount;
+                chunk_shifted[7] |= remainder_bits;
+            }
+            Some(Shift::Right(_, prev_remainder, curr_remainder)) => {
```

Comment:
`Shift::Right` is unconditionally initialized with `prev_remainder` and `curr_remainder` as `None`. Am I missing a place where either is set to `Some`? If not, this whole branch is a no-op.

Reply:
Yes, at the beginning of the for loop. The correct remainder needs to be extracted before shifting the bytes, so the extraction is performed before. Then, the remainder from the previous chunk of bytes is applied here